test/blackbox-tests/test-cases/actions/actions-stdxxx-on-success.t

```diff
@@ -0,0 +1,221 @@
+  > EOF
+
+  $ dune clean
+  $ dune build --actions-stdout-on-success=swallow
+            sh alias default
+  a.stderr
+            sh alias default
+  b.stderr
+
+Here, it isn't necessary to re-execute the second
+rule but we still do it:
+
+  $ dune build
+            sh alias default
+  a.stdout
+  a.stderr
+            sh alias default
+  b.stderr
+
+In case of errors
+-----------------
+
+In case of errors, we print everything no matter what.
+
+  $ cat > dune <<EOF
+  > (rule
+  >  (alias default)
+  >  (action (system "echo 'Hello, world!'; exit 1")))
+  > EOF
+
+  $ dune build
+  File "dune", line 1, characters 0-73:
+  1 | (rule
+  2 |  (alias default)
+  3 |  (action (system "echo 'Hello, world!'; exit 1")))
+            sh alias default (exit 1)
+  (cd _build/default && sh -c 'echo '\''Hello, world!'\''; exit 1')
+  Hello, world!
+  [1]
+
+  $ dune clean
+  $ dune build --actions-stdout=swallow
+  File "dune", line 1, characters 0-73:
+  1 | (rule
+  2 |  (alias default)
+  3 |  (action (system "echo 'Hello, world!'; exit 1")))
+            sh alias default (exit 1)
+  (cd _build/default && sh -c 'echo '\''Hello, world!'\''; exit 1')
+  Hello, world!
+  [1]
+
+  $ dune clean
+  $ dune build --actions-stdout=must-be-empty
+  File "dune", line 1, characters 0-73:
+  1 | (rule
+  2 |  (alias default)
+  3 |  (action (system "echo 'Hello, world!'; exit 1")))
+            sh alias default (exit 1)
+  (cd _build/default && sh -c 'echo '\''Hello, world!'\''; exit 1')
+  Hello, world!
+  [1]
+
+With compound actions
+---------------------
+
+At the moment, the behavior is a bit odd. We swallow the stdout of the
+first command but not the second:
+
+  $ cat > dune <<EOF
+  > (rule
+  >  (alias default)
+  >  (action
+  >   (progn
+  >    (system "echo 1")
+  >    (system "echo 2; exit 1"))))
+  > EOF
+
+  $ dune build --actions-stdout-on-success=swallow
+  File "dune", line 1, characters 0-93:
+  1 | (rule
+  2 |  (alias default)
+  3 |  (action
+  4 |   (progn
+  5 |    (system "echo 1")
+  6 |    (system "echo 2; exit 1"))))
+            sh alias default (exit 1)
+  (cd _build/default && sh -c 'echo 2; exit 1')
+  2
+  [1]
+
+For must-be-empty, if two programs print something without failing we
+stop at the first program. That's not terrible, but it would seem
+better if we stop at the end of the whole action.
+
+  $ cat > dune <<EOF
+  > (rule
+  >  (alias default)
+  >  (action
+  >   (progn
+  >    (system "echo 1")
+  >    (system "echo 2"))))
+  > EOF
+
+  $ dune build --actions-stdout-on-success=must-be-empty
+  File "dune", line 1, characters 0-85:
+  1 | (rule
+  2 |  (alias default)
+  3 |  (action
+  4 |   (progn
+  5 |    (system "echo 1")
+  6 |    (system "echo 2"))))
+            sh alias default (exit 0)
+  (cd _build/default && sh -c 'echo 1')
+  1
+  [1]
+
+With builtin actions
+--------------------
+
+We currently never swallow the output of builtin actions such as
+`echo`, which is odd:
```

Comment:
There are a few issues currently: not swallowing output of built-in actions, not treating compound actions correctly, etc. Perhaps, it's worth creating an issue, with a TODO list of all things that need to be fixed, so that we can track the progress there?

Reply:
Sounds good. All issues should be fixed by making these options operate on whole actions, as their name suggests. I created #4531. #4435 is related and exposes some other weirdness even without the new options.

Reply:
Great, thanks!